
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="shortcut icon" href="../../assets/images/dna-svgrepo-com.svg">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-6.0.2">
    
    
      
        <title>Variant calling - NGS - variant analysis</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.38780c08.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.3f72e892.min.css">
        
      
    
    
    
      
        
        <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto",-apple-system,BlinkMacSystemFont,Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono",SFMono-Regular,Consolas,Menlo,monospace}</style>
      
    
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
      
  
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#exercises" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid" aria-label="Header">
    <a href="../.." title="NGS - variant analysis" class="md-header-nav__button md-logo" aria-label="NGS - variant analysis">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M.1 494.1c-1.1 9.5 6.3 17.8 15.9 17.8l32.3.1c8.1 0 14.9-5.9 16-13.9.7-4.9 1.8-11.1 3.4-18.1H380c1.6 6.9 2.9 13.2 3.5 18.1 1.1 8 7.9 14 16 13.9l32.3-.1c9.6 0 17.1-8.3 15.9-17.8-4.6-37.9-25.6-129-118.9-207.7-17.6 12.4-37.1 24.2-58.5 35.4 6.2 4.6 11.4 9.4 17 14.2H159.7c21.3-18.1 47-35.6 78.7-51.4C410.5 199.1 442.1 65.8 447.9 17.9 449 8.4 441.6.1 432 .1L399.6 0c-8.1 0-14.9 5.9-16 13.9-.7 4.9-1.8 11.1-3.4 18.1H67.8c-1.6-7-2.7-13.1-3.4-18.1-1.1-8-7.9-14-16-13.9L16.1.1C6.5.1-1 8.4.1 17.9 5.3 60.8 31.4 171.8 160 256 31.5 340.2 5.3 451.2.1 494.1zM224 219.6c-25.1-13.7-46.4-28.4-64.3-43.6h128.5c-17.8 15.2-39.1 30-64.2 43.6zM355.1 96c-5.8 10.4-12.8 21.1-21 32H114c-8.3-10.9-15.3-21.6-21-32h262.1zM92.9 416c5.8-10.4 12.8-21.1 21-32h219.4c8.3 10.9 15.4 21.6 21.2 32H92.9z"/></svg>

    </a>
    <label class="md-header-nav__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header-nav__title" data-md-component="header-title">
      
        <div class="md-header-nav__ellipsis">
          <span class="md-header-nav__topic md-ellipsis">
            NGS - variant analysis
          </span>
          <span class="md-header-nav__topic md-ellipsis">
            
              Variant calling
            
          </span>
        </div>
      
    </div>
    
      <label class="md-header-nav__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0116 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 019.5 16 6.5 6.5 0 013 9.5 6.5 6.5 0 019.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active">
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0116 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 019.5 16 6.5 6.5 0 013 9.5 6.5 6.5 0 019.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" data-md-component="search-reset" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header-nav__source">
        
<a href="https://github.com/sib-swiss/NGS-variants-training" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/NGS-variants-training
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="NGS - variant analysis" class="md-nav__button md-logo" aria-label="NGS - variant analysis">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M.1 494.1c-1.1 9.5 6.3 17.8 15.9 17.8l32.3.1c8.1 0 14.9-5.9 16-13.9.7-4.9 1.8-11.1 3.4-18.1H380c1.6 6.9 2.9 13.2 3.5 18.1 1.1 8 7.9 14 16 13.9l32.3-.1c9.6 0 17.1-8.3 15.9-17.8-4.6-37.9-25.6-129-118.9-207.7-17.6 12.4-37.1 24.2-58.5 35.4 6.2 4.6 11.4 9.4 17 14.2H159.7c21.3-18.1 47-35.6 78.7-51.4C410.5 199.1 442.1 65.8 447.9 17.9 449 8.4 441.6.1 432 .1L399.6 0c-8.1 0-14.9 5.9-16 13.9-.7 4.9-1.8 11.1-3.4 18.1H67.8c-1.6-7-2.7-13.1-3.4-18.1-1.1-8-7.9-14-16-13.9L16.1.1C6.5.1-1 8.4.1 17.9 5.3 60.8 31.4 171.8 160 256 31.5 340.2 5.3 451.2.1 494.1zM224 219.6c-25.1-13.7-46.4-28.4-64.3-43.6h128.5c-17.8 15.2-39.1 30-64.2 43.6zM355.1 96c-5.8 10.4-12.8 21.1-21 32H114c-8.3-10.9-15.3-21.6-21-32h262.1zM92.9 416c5.8-10.4 12.8-21.1 21-32h219.4c8.3 10.9 15.4 21.6 21.2 32H92.9z"/></svg>

    </a>
    NGS - variant analysis
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/sib-swiss/NGS-variants-training" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/NGS-variants-training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href="../.." class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../../precourse/" class="md-nav__link">
      Precourse preparations
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../../course_schedule/" class="md-nav__link">
      Course schedule
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-nav__toggle md-toggle" data-md-toggle="nav-4" type="checkbox" id="nav-4">
    
    <label class="md-nav__link" for="nav-4">
      Day 1
      <span class="md-nav__icon md-icon"></span>
    </label>
    <nav class="md-nav" aria-label="Day 1" data-md-level="1">
      <label class="md-nav__title" for="nav-4">
        <span class="md-nav__icon md-icon"></span>
        Day 1
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../../day1/introduction/" class="md-nav__link">
      Introduction
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../day1/server_login/" class="md-nav__link">
      Server login
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../day1/alignment/" class="md-nav__link">
      Read alignment - basics
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../day1/alignment_advanced/" class="md-nav__link">
      Read alignment - advanced
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-nav__toggle md-toggle" data-md-toggle="nav-5" type="checkbox" id="nav-5" checked>
    
    <label class="md-nav__link" for="nav-5">
      Day 2
      <span class="md-nav__icon md-icon"></span>
    </label>
    <nav class="md-nav" aria-label="Day 2" data-md-level="1">
      <label class="md-nav__title" for="nav-5">
        <span class="md-nav__icon md-icon"></span>
        Day 2
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        Variant calling
        <span class="md-nav__icon md-icon"></span>
      </label>
    
    <a href="./" class="md-nav__link md-nav__link--active">
      Variant calling
    </a>
    
      
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    Exercises
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-indexing-indexing-indexing" class="md-nav__link">
    1. Indexing, indexing, indexing
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-base-quality-score-recalibration-bqsr" class="md-nav__link">
    2. Base Quality Score Recalibration (BQSR)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-variant-calling" class="md-nav__link">
    3. Variant calling
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-visualisation" class="md-nav__link">
    4. Visualisation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-combining-gvcfs" class="md-nav__link">
    5. Combining GVCFs
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
    
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../filtering_evaluation/" class="md-nav__link">
      Filtering & evaluation
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    Exercises
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-indexing-indexing-indexing" class="md-nav__link">
    1. Indexing, indexing, indexing
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-base-quality-score-recalibration-bqsr" class="md-nav__link">
    2. Base Quality Score Recalibration (BQSR)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-variant-calling" class="md-nav__link">
    3. Variant calling
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-visualisation" class="md-nav__link">
    4. Visualisation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-combining-gvcfs" class="md-nav__link">
    5. Combining GVCFs
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                
                  <h1>Variant calling</h1>
                
                <h2 id="exercises">Exercises</h2>
<h3 id="1-indexing-indexing-indexing">1. Indexing, indexing, indexing</h3>
<p>Many algorithms work faster, or only work with an index of their (large) input files. In that sense, <code>gatk</code> is no different from other tools. The index for a reference needs to be created in two steps:</p>
<div class="highlight"><pre><span></span><code><span class="nb">cd</span> ~/data/reference
samtools faidx &lt;reference.fa&gt;
gatk CreateSequenceDictionary --REFERENCE &lt;reference.fa&gt;
</code></pre></div>
<p>Also input vcf files need to be indexed. This will create a <code>.idx</code> file associated with the <code>.vcf</code>. You can do this like this:</p>
<div class="highlight"><pre><span></span><code>gatk IndexFeatureFile --input &lt;variants.vcf&gt;
</code></pre></div>
<p><strong>Exercise:</strong> Create the required <code>gatk</code> indexes for:</p>
<ul>
<li>The reference genome <code>reference/Homo_sapiens.GRCh38.dna.chromosome.20.fa</code></li>
<li>A part of the dbsnp database: <code>variants/GCF.38.filtered.renamed.vcf</code></li>
<li>A part of the 1000 genomes indel golden standard: <code>variants/1000g_gold_standard.indels.filtered.vcf</code></li>
</ul>
<details class="done"><summary>Answer</summary><p>Creating the index for the reference genome:</p>
<div class="highlight"><pre><span></span><code><span class="nb">cd</span> ~/data
samtools faidx reference/Homo_sapiens.GRCh38.dna.chromosome.20.fa
gatk CreateSequenceDictionary --REFERENCE reference/Homo_sapiens.GRCh38.dna.chromosome.20.fa
</code></pre></div>
<p>Creating the indices for the vcfs:</p>
<div class="highlight"><pre><span></span><code>gatk IndexFeatureFile --input variants/1000g_gold_standard.indels.filtered.vcf
gatk IndexFeatureFile --input variants/GCF.38.filtered.renamed.vcf
</code></pre></div>
</details>
<div class="admonition note">
<p class="admonition-title">Chromosome names</p>
<p>Unlike IGV, <code>gatk</code> requires equal chromosome names for all its input files and indexes, e.g. in <code>.fasta</code>, <code>.bam</code> and <code>.vcf</code> files. In general, for the human genome there are three types of chromosome names:</p>
<ul>
<li>Just a number, e.g. <code>20</code></li>
<li>Prefixed by <code>chr</code>. e.g. <code>chr20</code></li>
<li>Refseq name, e.g. <code>NC_000020.11</code></li>
</ul>
<p>Before you start the alignment, it&rsquo;s wise to check out what chromosome naming your input files are using, because changing chromosome names in a <code>.fasta</code> file is easier than in a <code>.bam</code> file.</p>
<p>If your fasta titles are e.g. starting with a number you can add <code>chr</code> to it with <code>sed</code>:</p>
<div class="highlight"><pre><span></span><code>sed s/^&gt;/&gt;chr/g &lt;reference.fasta&gt;
</code></pre></div>
<p>You can change chromsome names in a vcf with <a href="http://samtools.github.io/bcftools/bcftools.html#annotate"><code>bcftools annotate</code></a>:</p>
<div class="highlight"><pre><span></span><code>bcftools annotate --rename-chrs &lt;tab-delimited-renaming&gt; &lt;input.vcf&gt;
</code></pre></div>
</div>
<h3 id="2-base-quality-score-recalibration-bqsr">2. Base Quality Score Recalibration (BQSR)</h3>
<p>BQSR evaluates the base qualities on systematic error. It can ignore sites with known variants. BQSR helps to identify faulty base calls, and therefore reduces the chance on discovering false positive variant positions.</p>
<p>BQSR is done in two steps:</p>
<ol>
<li>Recalibration with <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037593511-BaseRecalibrator"><code>gatk BaseRecalibrator</code></a></li>
<li>By using the output of <code>gatk BaseRecalibrator</code>, the application to the bam file with <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037055712-ApplyBQSR"><code>gatk ApplyBQSR</code></a></li>
</ol>
<p><strong>Exercise:</strong> Check out the documentation of the tools. Which options are required?</p>
<details class="done"><summary>Answer</summary><p>For <code>gatk BaseRecalibrator</code>:</p>
<ul>
<li><code>--reference</code></li>
<li><code>--input</code></li>
<li><code>--known-sites</code></li>
<li><code>--output</code></li>
</ul>
<p>For <code>gatk ApplyBQSR</code>:</p>
<ul>
<li><code>--bqsr-recal-file</code></li>
<li><code>--input</code></li>
<li><code>--output</code></li>
</ul>
</details>
<p><strong>Exercise:</strong> Run the two commands with the required options on <code>mother.bam</code>, with <code>--known-sites</code> <code>variants/1000g_gold_standard.indels.filtered.vcf</code> and <code>variants/GCF.38.filtered.renamed.vcf</code>.</p>
<div class="admonition hint">
<p class="admonition-title">Multiple inputs for same argument</p>
<p>In some cases you need to add multiple inputs (e.g. multiple <code>vcf</code> files) into the same argument (e.g. <code>--known-sites</code>). To provide multiple inputs for the same argument in <code>gatk</code>, you can use the same argument multiple times, e.g.:</p>
<div class="highlight"><pre><span></span><code>gatk BaseRecalibrator <span class="se">\</span>
--reference &lt;reference.fa&gt; <span class="se">\</span>
--input &lt;alignment.bam&gt; <span class="se">\</span>
--known-sites &lt;variants1.vcf&gt; <span class="se">\</span>
--known-sites &lt;variants2.vcf&gt; <span class="se">\</span>
--output &lt;output.table&gt;
</code></pre></div>
</div>
<details class="done"><summary>Answer</summary><div class="highlight"><pre><span></span><code><span class="nb">cd</span> ~/data

mkdir bqsr

gatk BaseRecalibrator <span class="se">\</span>
--reference reference/Homo_sapiens.GRCh38.dna.chromosome.20.fa <span class="se">\</span>
--input alignment/mother.bam <span class="se">\</span>
--known-sites variants/GCF.38.filtered.renamed.vcf <span class="se">\</span>
--known-sites variants/1000g_gold_standard.indels.filtered.vcf <span class="se">\</span>
--output bqsr/mother.recal.table

gatk ApplyBQSR <span class="se">\</span>
--input alignment/mother.bam <span class="se">\</span>
--bqsr-recal-file bqsr/mother.recal.table <span class="se">\</span>
--output bqsr/mother.recal.bam
</code></pre></div>
</details>
<p><strong>Exercise:</strong> Place these commands in a loop, that performs the BQSR for mother, father and son.</p>
<details class="done"><summary>Answer</summary><div class="highlight"><pre><span></span><code><span class="nb">cd</span> ~/data

<span class="k">for</span> sample in mother father son
<span class="k">do</span>
  gatk BaseRecalibrator <span class="se">\</span>
  --reference reference/Homo_sapiens.GRCh38.dna.chromosome.20.fa <span class="se">\</span>
  --input alignment/<span class="nv">$sample</span>.bam <span class="se">\</span>
  --known-sites variants/GCF.38.filtered.renamed.vcf <span class="se">\</span>
  --known-sites variants/1000g_gold_standard.indels.filtered.vcf <span class="se">\</span>
  --output bqsr/<span class="nv">$sample</span>.recal.table

  gatk ApplyBQSR <span class="se">\</span>
  --input alignment/<span class="nv">$sample</span>.bam <span class="se">\</span>
  --bqsr-recal-file bqsr/<span class="nv">$sample</span>.recal.table <span class="se">\</span>
  --output bqsr/<span class="nv">$sample</span>.recal.bam
<span class="k">done</span>
</code></pre></div>
</details>
<h3 id="3-variant-calling">3. Variant calling</h3>
<p>The command <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037225632-HaplotypeCaller"><code>gatk HaplotypeCaller</code></a> is the core command of <code>gatk</code>. It performs the actual variant calling.</p>
<p><strong>Exercise:</strong> Check out the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037225632-HaplotypeCaller"><code>gatk HaplotypeCaller</code> documentation</a>, and find out which arguments are required.</p>
<details class="done"><summary>Answer</summary><p>Required arguments are:</p>
<ul>
<li><code>--input</code></li>
<li><code>--ouput</code></li>
<li><code>--reference</code></li>
</ul>
</details>
<p><strong>Exercise:</strong> Run <code>gatk HaplotypeCaller</code> with required options on the recalibrated alignment file of the mother (<code>bqsr/mother.recal.bam</code>). We&rsquo;ll focus on a small region, so add <code>--intervals chr20:10018000-10220000</code>.</p>
<details class="done"><summary>Answer</summary><div class="highlight"><pre><span></span><code>gatk HaplotypeCaller <span class="se">\</span>
--reference reference/Homo_sapiens.GRCh38.dna.chromosome.20.fa <span class="se">\</span>
--input bqsr/mother.recal.bam <span class="se">\</span>
--output variants/mother.HC.vcf <span class="se">\</span>
--intervals chr20:10018000-10220000 <span class="se">\</span>
</code></pre></div>
</details>
<p>We will do the variant calling on all three samples. Later we want to combine the variant calls. For efficient merging of vcfs, we will need to output the variants as a GVCF. To do that, we will use the option <code>--emit-ref-confidence GVCF</code>. Also, we&rsquo;ll visualise the haplotype phasing with IGV in the next section. For that we&rsquo;ll need a phased bam. You can get this output with the argument <code>--bam-output</code>.</p>
<p><strong>Exercise:</strong> Run <code>gatk HaplotypeCaller</code> for mother, father and son by using a loop, and by using the arguments in the previous exercise. On top of that add the arguments <code>--emit-ref-confidence GVCF</code> and <code>--bamoutput &lt;phased.bam&gt;</code>.</p>
<details class="done"><summary>Answer</summary><div class="highlight"><pre><span></span><code><span class="k">for</span> sample in mother father son
<span class="k">do</span>
  gatk HaplotypeCaller <span class="se">\</span>
  --reference reference/Homo_sapiens.GRCh38.dna.chromosome.20.fa <span class="se">\</span>
  --input bqsr/<span class="nv">$sample</span>.recal.bam <span class="se">\</span>
  --output variants/<span class="nv">$sample</span>.HC.g.vcf <span class="se">\</span>
  --bam-output variants/<span class="nv">$sample</span>.phased.bam <span class="se">\</span>
  --intervals chr20:10018000-10220000 <span class="se">\</span>
  --emit-ref-confidence GVCF
<span class="k">done</span>
</code></pre></div>
</details>
<h3 id="4-visualisation">4. Visualisation</h3>
<p>Download the following files to your local computer:</p>
<ul>
<li><code>variants/mother.phased.bam</code></li>
<li><code>variants/mother.HC.g.vcf</code></li>
<li><code>bqsr/mother.recal.bam</code></li>
</ul>
<p>Launch IGV and select the human genome version hg38 as a reference.</p>
<figure>
  <img src="../../assets/images/select_hg38.png" width="300"/>
</figure>

<p>Load the downloaded files as tracks in igv with <strong>File &gt; Load From File&hellip;</strong>, and navigate to region <code>chr20:10,026,397-10,026,638</code>.</p>
<p><strong>Exercise:</strong> Zoom out for a bit. Not all reads are in the track of <code>mother.phased.bam</code>. What kind of reads are in there?</p>
<details class="done"><summary>Answer</summary><p>The reads supporting called variants.</p>
</details>
<p>Now, we&rsquo;ll investigate the haplotype phasing.  Go back to <code>chr20:10,026,397-10,026,638</code>.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>If your screen isn&rsquo;t huge, you can remove the track <code>mother.recal.bam</code>. Do that by right-click on the track, and click on <strong>Remove Track</strong>.</p>
</div>
<p>In the track with <code>mother.phased.bam</code>, right click on the reads and select <strong>Group alignments by &gt; read group</strong>. This splits your track in two parts, one with artificial reads describing haplotypes that were taken in consideration (ArtificalHaplotypeRG), and one with original reads that support the variants.</p>
<p><strong>Exercise</strong>: How many haplotypes were taken into consideration? How many haplotypes can you expect at maximum within a single individual?</p>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>This might come as a shock, but humans are diploid.</p>
</div>
<details class="done"><summary>Answer</summary><p>Three haplotypes, as there are three artificial reads:</p>
<p><figure>
  <img src="../../assets/images/artificial_reads.png" width="500"/>
</figure></p>
<p>Diploids can carry two haplotypes. So at least one of the three is wrong.</p>
</details>
<p>Now colour the reads by phase. Do that with by right clicking on the track and choose <strong>Colour alignments by &gt; tag</strong>, and type in &ldquo;HC&rdquo; (that&rsquo;s the tag where the phasing is stored).</p>
<p><strong>Exercise:</strong> Which artificial read doesn&rsquo;t get support from the original sequence reads? Are the alternative alleles of the two SNPs on the same haplotype (i.e. in phase)?</p>
<details class="done"><summary>Answer</summary><p>The track should look like this (colours can be different):</p>
<p><figure>
  <img src="../../assets/images/phased_bam.png" width="500"/>
</figure></p>
<p>The reads only support the brown and blue haplotype, and not the pink one.</p>
<p>The alternative alleles are coloured in IGV. For the first SNP this is the C (in blue) and for the second the T (in red). They are always in different reads, so they are in repulsion (not in phase).</p>
</details>
<h3 id="5-combining-gvcfs">5. Combining GVCFs</h3>
<p>Now that we have all three GVCFs of the mother, father and son, we can combine them into a database. We do this because it enables us to later add GVCFs (with the option <code>--genomicsdb-update-workspace-path</code>), and to efficiently combine them into a single vcf.</p>
<p>You can generate a GenomicsDB on our three samples like this:</p>
<div class="highlight"><pre><span></span><code>gatk GenomicsDBImport <span class="se">\</span>
--variant variants/mother.HC.g.vcf <span class="se">\</span>
--variant variants/father.HC.g.vcf <span class="se">\</span>
--variant variants/son.HC.g.vcf <span class="se">\</span>
--intervals chr20:10018000-10220000 <span class="se">\</span>
--genomicsdb-workspace-path genomicsdb
</code></pre></div>
<p><strong>Exercise:</strong> Run this command to generate the database.</p>
<p>You can retrieve the combined vcf from the database with <code>gatk GenotypeGVCFs</code>.</p>
<div class="highlight"><pre><span></span><code>gatk GenotypeGVCFs <span class="se">\</span>
--reference reference/Homo_sapiens.GRCh38.dna.chromosome.20.fa <span class="se">\</span>
--variant gendb://genomicsdb <span class="se">\</span>
--intervals chr20:10018000-10220000 <span class="se">\</span>
--output variants/trio.vcf
</code></pre></div>
<p><strong>Exercise:</strong> Run this command to generate the combined vcf.</p>
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid" aria-label="Footer">
        
          <a href="../../day1/alignment_advanced/" class="md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
            </div>
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Read alignment - advanced
              </div>
            </div>
          </a>
        
        
          <a href="../filtering_evaluation/" class="md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Filtering & evaluation
              </div>
            </div>
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../assets/javascripts/vendor.77e55a48.min.js"></script>
      <script src="../../assets/javascripts/bundle.9554a270.min.js"></script><script id="__lang" type="application/json">{"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}</script>
      
      <script>
        app = initialize({
          base: "../..",
          features: [],
          search: Object.assign({
            worker: "../../assets/javascripts/worker/search.4ac00218.min.js"
          }, typeof search !== "undefined" && search)
        })
      </script>
      
    
  </body>
</html>