

<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://sib-swiss.github.io/NGS-variants-training/2022.9/day2/filtering_evaluation/">
      
      <link rel="icon" href="../../assets/images/SIB_logo.svg">
      <meta name="generator" content="mkdocs-1.3.0, mkdocs-material-8.2.8">
    
    
      
        <title>Filtering & evaluation - NGS - variant analysis</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.644de097.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.e6a45f82.min.css">
        
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-SK9D4D2P4M"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-SK9D4D2P4M');
</script>

    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#learning-outcomes" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-component="outdated" hidden>
        <aside class="md-banner md-banner--warning">
          
        </aside>
      </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="NGS - variant analysis" class="md-header__button md-logo" aria-label="NGS - variant analysis" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            NGS - variant analysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Filtering & evaluation
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/sib-swiss/NGS-variants-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/NGS-variants-training
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="NGS - variant analysis" class="md-nav__button md-logo" aria-label="NGS - variant analysis" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    NGS - variant analysis
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/sib-swiss/NGS-variants-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/NGS-variants-training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../precourse/" class="md-nav__link">
        Precourse preparations
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../course_schedule/" class="md-nav__link">
        Course schedule
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4">
          Day 1
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Day 1" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Day 1
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/introduction/" class="md-nav__link">
        Introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/server_login/" class="md-nav__link">
        Setup
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/reproducibility/" class="md-nav__link">
        Reproducibility
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/alignment/" class="md-nav__link">
        Read alignment - basics
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/alignment_advanced/" class="md-nav__link">
        Read alignment - advanced
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_5">
          Day 2
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Day 2" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          Day 2
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../variant_calling_preparation/" class="md-nav__link">
        Variant calling - preparation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../variant_calling/" class="md-nav__link">
        Variant calling
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../visualisation/" class="md-nav__link">
        Visualisation
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Filtering & evaluation
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Filtering & evaluation
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    Learning outcomes
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    Material
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    Exercises
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-hard-filtering" class="md-nav__link">
    1. Hard filtering
  </a>
  
    <nav class="md-nav" aria-label="1. Hard filtering">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#splitting-snps-and-indels" class="md-nav__link">
    Splitting SNPs and INDELs
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering-snps" class="md-nav__link">
    Filtering SNPs
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering-indels" class="md-nav__link">
    Filtering INDELs
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#merging-filtered-snps-and-indels" class="md-nav__link">
    Merging filtered SNPs and INDELs
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-evaluation-by-concordance" class="md-nav__link">
    2. Evaluation by concordance
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../annotation/" class="md-nav__link">
        Annotation
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    Learning outcomes
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    Material
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    Exercises
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-hard-filtering" class="md-nav__link">
    1. Hard filtering
  </a>
  
    <nav class="md-nav" aria-label="1. Hard filtering">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#splitting-snps-and-indels" class="md-nav__link">
    Splitting SNPs and INDELs
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering-snps" class="md-nav__link">
    Filtering SNPs
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering-indels" class="md-nav__link">
    Filtering INDELs
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#merging-filtered-snps-and-indels" class="md-nav__link">
    Merging filtered SNPs and INDELs
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-evaluation-by-concordance" class="md-nav__link">
    2. Evaluation by concordance
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
<a href="https://github.com/sib-swiss/NGS-variants-training/edit/main/docs/day2/filtering_evaluation.md" title="Edit this page" class="md-content__button md-icon">
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
</a>



  <h1>Filtering & evaluation</h1>

<h2 id="learning-outcomes">Learning outcomes</h2>
<p><strong>After having completed this chapter you will be able to:</strong></p>
<ul>
<li>Explain why using Variant Quality Score Recalibration (VQSR) for filtering variants can outperform hard filtering </li>
<li>Perform hard filtering on both SNPs and INDELs separately by using <code>gatk SelectVariants</code> in combination with <code>gatk VariantFiltration</code></li>
<li>Perform concordance between called variants and a truth set and evaluate performance of a variant calling workflow</li>
</ul>
<h2 id="material">Material</h2>
<p><a class="md-button" href="../../assets/pdf/filtering_evaluation.pdf"><span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M88 304h-8v-48h8c13.3 0 24 10.7 24 24s-10.7 24-24 24zm104-48h8c8.8 0 16 7.2 16 16v64c0 8.8-7.2 16-16 16h-8v-96zM224 0v128c0 17.7 14.3 32 32 32h128v288c0 35.3-28.7 64-64 64H64c-35.35 0-64-28.7-64-64V64C0 28.65 28.65 0 64 0h160zM64 224c-8.84 0-16 7.2-16 16v128c0 8.8 7.16 16 16 16s16-7.2 16-16v-32h8c30.9 0 56-25.1 56-56s-25.1-56-56-56H64zm96 144c0 8.8 7.2 16 16 16h24c26.5 0 48-21.5 48-48v-64c0-26.5-21.5-48-48-48h-24c-8.8 0-16 7.2-16 16v128zm128-144c-8.8 0-16 7.2-16 16v128c0 8.8 7.2 16 16 16s16-7.2 16-16v-48h32c8.8 0 16-7.2 16-16s-7.2-16-16-16h-32v-32h32c8.8 0 16-7.2 16-16s-7.2-16-16-16h-48zM256 0l128 128H256V0z"/></svg></span> Download the presentation</a></p>
<h2 id="exercises">Exercises</h2>
<h3 id="1-hard-filtering">1. Hard filtering</h3>
<p>The developers of <code>gatk</code> strongly advise to do <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531612-Variant-Quality-Score-Recalibration-VQSR-">Variant Quality Score Recalibration (VQSR)</a> for filtering SNPs and INDELs. However, this is not always possible. For example, in the case of limited data availability and/or in the case you are working with non-model organisms and/or in the case you are a bit lazy and okay with a number of false positives.</p>
<p>Our dataset is too small to apply VQSR. We will therefore do hard filtering instead.</p>
<h4 id="splitting-snps-and-indels">Splitting SNPs and INDELs</h4>
<p>First, filtering thresholds are usually different for SNPs and INDELs. Therefore, we will split <code>trio.vcf</code> into two vcfs, one containg only SNPs, and one containing only INDELs. You can extract all the SNP records in our trio vcf like this:</p>
<div class="highlight"><pre><span></span><code><span class="nb">cd</span> ~/workdir

gatk SelectVariants <span class="se">\</span>
--variant variants/trio.vcf <span class="se">\</span>
--select-type-to-include SNP <span class="se">\</span>
--output variants/trio.SNP.vcf
</code></pre></div>
<p><strong>Exercise:</strong> Check out the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037055952-SelectVariants">documentation of <code>gatk SelectVariants</code></a>, and:</p>
<ul>
<li>Figure out what you&rsquo;ll need to write at <code>--select-type-to-include</code> if you want to select only INDELS.</li>
<li>Make a script (named <code>C09_select_SNPs.sh</code>) to generate a vcf with only the SNPs</li>
<li>Make a script (named <code>C10_select_INDELs.sh</code>) to generate a second vcf with only the INDELs from <code>trio.vcf</code>.</li>
</ul>
<details class="done"><summary>Answer</summary><p>You will need to fill in <code>INDEL</code> at <code>--select-type-to-include</code> to filter for INDELs.</p>
<p>To get the SNPs you can run the command above:</p>
<div class="highlight"><span class="filename">C09_select_SNPs.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span> ~/workdir

gatk SelectVariants <span class="se">\</span>
--variant results/variants/trio.vcf <span class="se">\</span>
--select-type-to-include SNP <span class="se">\</span>
--output results/variants/trio.SNP.vcf
</code></pre></div>
<p>To get the INDELs you&rsquo;ll need to change <code>--select-type-to-include</code> to <code>INDEL</code>:</p>
<div class="highlight"><span class="filename">C10_select_INDELs.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span> ~/workdir

gatk SelectVariants <span class="se">\</span>
--variant results/variants/trio.vcf <span class="se">\</span>
--select-type-to-include INDEL <span class="se">\</span>
--output results/variants/trio.INDEL.vcf
</code></pre></div>
</details>
<h4 id="filtering-snps">Filtering SNPs</h4>
<p>The command <code>gatk VariantFiltration</code> enables you to filter for both the INFO field (per variant) and FORMAT field (per genotype). For now we&rsquo;re only interested in filtering variants. Below you can find the command to hard-filter the SNP variants on some sensible thresholds (that are explained <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890471-Hard-filtering-germline-short-variants">here</a>).</p>
<div class="highlight"><pre><span></span><code>gatk VariantFiltration <span class="se">\</span>
--variant variants/trio.SNP.vcf <span class="se">\</span>
--filter-expression <span class="s2">&quot;QD &lt; 2.0&quot;</span>              --filter-name <span class="s2">&quot;QD2&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;QUAL &lt; 30.0&quot;</span>           --filter-name <span class="s2">&quot;QUAL30&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;SOR &gt; 3.0&quot;</span>             --filter-name <span class="s2">&quot;SOR3&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;FS &gt; 60.0&quot;</span>             --filter-name <span class="s2">&quot;FS60&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;MQ &lt; 40.0&quot;</span>             --filter-name <span class="s2">&quot;MQ40&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;MQRankSum &lt; -12.5&quot;</span>     --filter-name <span class="s2">&quot;MQRankSum-12.5&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;ReadPosRankSum &lt; -8.0&quot;</span> --filter-name <span class="s2">&quot;ReadPosRankSum-8&quot;</span> <span class="se">\</span>
--output variants/trio.SNP.filtered.vcf
</code></pre></div>
<p><strong>Exercise:</strong> Run the filtering command above in a script called <code>C11_filter_SNPs.sh</code>. Did it affect the number of records in the vcf?</p>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>You can check out the number of records in a vcf with:</p>
<div class="highlight"><pre><span></span><code>grep -v <span class="s2">&quot;^#&quot;</span> &lt;variants.vcf&gt; <span class="p">|</span> wc -l
</code></pre></div>
</div>
<details class="done"><summary>Answer</summary><p>Your script:</p>
<div class="highlight"><span class="filename">C11_filter_SNPs.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span> ~/workdir/results/variants

gatk VariantFiltration <span class="se">\</span>
--variant trio.SNP.vcf <span class="se">\</span>
--filter-expression <span class="s2">&quot;QD &lt; 2.0&quot;</span>              --filter-name <span class="s2">&quot;QD2&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;QUAL &lt; 30.0&quot;</span>           --filter-name <span class="s2">&quot;QUAL30&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;SOR &gt; 3.0&quot;</span>             --filter-name <span class="s2">&quot;SOR3&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;FS &gt; 60.0&quot;</span>             --filter-name <span class="s2">&quot;FS60&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;MQ &lt; 40.0&quot;</span>             --filter-name <span class="s2">&quot;MQ40&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;MQRankSum &lt; -12.5&quot;</span>     --filter-name <span class="s2">&quot;MQRankSum-12.5&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;ReadPosRankSum &lt; -8.0&quot;</span> --filter-name <span class="s2">&quot;ReadPosRankSum-8&quot;</span> <span class="se">\</span>
--output trio.SNP.filtered.vcf
</code></pre></div>
<p>There are no differences in the number of records:</p>
<div class="highlight"><pre><span></span><code>grep -v <span class="s2">&quot;^#&quot;</span> variants/trio.SNP.vcf <span class="p">|</span> wc -l
</code></pre></div>
<p>and</p>
<div class="highlight"><pre><span></span><code>grep -v <span class="s2">&quot;^#&quot;</span> variants/trio.SNP.filtered.vcf <span class="p">|</span> wc -l
</code></pre></div>
<p>both give 446.</p>
<p>However, there are SNPs filtered out, by changing the <code>FILTER</code> column. You can check the number of records with PASS by:</p>
<div class="highlight"><pre><span></span><code>grep -v <span class="s2">&quot;^#&quot;</span> variants/trio.SNP.filtered.vcf <span class="p">|</span> cut -f <span class="m">7</span> <span class="p">|</span> sort <span class="p">|</span> uniq -c
</code></pre></div>
<p>Giving:</p>
<div class="highlight"><pre><span></span><code>441 PASS
2 QD2;SOR3
3 SOR3
</code></pre></div>
</details>
<h4 id="filtering-indels">Filtering INDELs</h4>
<p>A command with <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531112--How-to-Filter-variants-either-with-VQSR-or-by-hard-filtering">sensible parameters</a> to do a first iteration of hard filtering the INDELs would be:</p>
<div class="highlight"><pre><span></span><code>gatk VariantFiltration <span class="se">\</span>
--variant variants/trio.INDEL.vcf <span class="se">\</span>
--filter-expression <span class="s2">&quot;QD &lt; 2.0&quot;</span>                  --filter-name <span class="s2">&quot;QD2&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;QUAL &lt; 30.0&quot;</span>               --filter-name <span class="s2">&quot;QUAL30&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;FS &gt; 200.0&quot;</span>                --filter-name <span class="s2">&quot;FS200&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;ReadPosRankSum &lt; -20.0&quot;</span>    --filter-name <span class="s2">&quot;ReadPosRankSum-20&quot;</span> <span class="se">\</span>
--output variants/trio.INDEL.filtered.vcf
</code></pre></div>
<p><strong>Exercise:</strong> Run the command from a script called <code>C12_filter_INDELs.sh</code> and figure out how many variants are filtered out.</p>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>You can use this command from the answer to the previous exercise:</p>
<div class="highlight"><pre><span></span><code>grep -v <span class="s2">&quot;^#&quot;</span> &lt;variants.vcf&gt; <span class="p">|</span> cut -f <span class="m">7</span> <span class="p">|</span> sort <span class="p">|</span> uniq -c
</code></pre></div>
<p>to see how many INDELs were filtered out.</p>
</div>
<details class="done"><summary>Answer</summary><p>Your script:</p>
<div class="highlight"><span class="filename">C12_filter_INDELs.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span> ~/workdir/results/variants

gatk VariantFiltration <span class="se">\</span>
--variant trio.INDEL.vcf <span class="se">\</span>
--filter-expression <span class="s2">&quot;QD &lt; 2.0&quot;</span>                  --filter-name <span class="s2">&quot;QD2&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;QUAL &lt; 30.0&quot;</span>               --filter-name <span class="s2">&quot;QUAL30&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;FS &gt; 200.0&quot;</span>                --filter-name <span class="s2">&quot;FS200&quot;</span> <span class="se">\</span>
--filter-expression <span class="s2">&quot;ReadPosRankSum &lt; -20.0&quot;</span>    --filter-name <span class="s2">&quot;ReadPosRankSum-20&quot;</span> <span class="se">\</span>
--output trio.INDEL.filtered.vcf
</code></pre></div>
<p>And check out the contents of the <code>FILTER</code> column: </p>
<div class="highlight"><pre><span></span><code>grep -v <span class="s2">&quot;^#&quot;</span> variants/trio.INDEL.filtered.vcf <span class="p">|</span> cut -f <span class="m">7</span> <span class="p">|</span> sort <span class="p">|</span> uniq -c
</code></pre></div>
<p>gives:</p>
<div class="highlight"><pre><span></span><code>110 PASS
</code></pre></div>
<p>So no variants are filtered out.</p>
</details>
<h4 id="merging-filtered-snps-and-indels">Merging filtered SNPs and INDELs</h4>
<p>Now that we have filtered the INDELs and SNPs separately, we can merge them again with this command:</p>
<div class="highlight"><pre><span></span><code>gatk MergeVcfs <span class="se">\</span>
--INPUT &lt;input1.vcf&gt; <span class="se">\</span>
--INPUT &lt;input2.vcf&gt; <span class="se">\</span>
--OUTPUT &lt;merged.vcf&gt;
</code></pre></div>
<p><strong>Exercise:</strong> Run this command from a script called <code>C13_merge_filtered.sh</code> to merge the vcfs (<code>trio.SNP.filtered.vcf</code> and <code>trio.INDEL.filtered.vcf</code>).</p>
<details class="done"><summary>Answer</summary><div class="highlight"><span class="filename">C13_merged_filtered.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span> ~/workdir/results/variants

gatk MergeVcfs <span class="se">\</span>
--INPUT trio.SNP.filtered.vcf <span class="se">\</span>
--INPUT trio.INDEL.filtered.vcf <span class="se">\</span>
--OUTPUT trio.filtered.vcf
</code></pre></div>
</details>
<h3 id="2-evaluation-by-concordance">2. Evaluation by concordance</h3>
<p>For this region we have a highly curated truth set for the mother available. It originates from the <a href="https://www.illumina.com/platinumgenomes.html">Illumina Platinum truth set</a>. You can find it at <code>data/variants/NA12878.vcf.gz</code></p>
<p>To check how well we did, we&rsquo;d first need to extract a vcf with only the information of the mother.</p>
<p><strong>Exercise:</strong> Generate a script called <code>C14_extract_mother_only.sh</code> to extract variants that have at least one alternative allele in the mother from <code>variants/trio.filtered.vcf</code>. Use <code>gatk SelectVariants</code> with the arguments:</p>
<ul>
<li><code>--sample-name mother</code></li>
<li><code>--exclude-non-variants</code></li>
<li><code>--remove-unused-alternates</code></li>
</ul>
<p>In addition to the required arguments.</p>
<details class="done"><summary>Answer</summary><div class="highlight"><span class="filename">C14_extract_mother_only.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span> ~/workdir/results/variants

gatk SelectVariants <span class="se">\</span>
--variant trio.filtered.vcf <span class="se">\</span>
--sample-name mother <span class="se">\</span>
--exclude-non-variants <span class="se">\</span>
--remove-unused-alternates <span class="se">\</span>
--output mother.trio.filtered.vcf
</code></pre></div>
</details>
<p><strong>Exercise:</strong></p>
<p>A. How many variants are in the extracted vcf? How many of those are filtered out?</p>
<p>B. Compare our vcf with the curated truth set with the command below from a script called <code>C15_evaluate_concordance.sh</code>. How many SNPs didn&rsquo;t we detect?</p>
<div class="highlight"><pre><span></span><code>gatk Concordance <span class="se">\</span>
--evaluation variants/mother.trio.filtered.vcf <span class="se">\</span>
--truth data/variants/NA12878.vcf.gz <span class="se">\</span>
--intervals chr20:10018000-10220000 <span class="se">\</span>
--summary variants/concordance.mother.trio.filtered
</code></pre></div>
<details class="done"><summary>Answer</summary><p>To get the number of records per FILTER, we run:</p>
<div class="highlight"><pre><span></span><code>grep -v <span class="s2">&quot;^#&quot;</span> variants/mother.trio.filtered.vcf <span class="p">|</span> cut -f <span class="m">7</span> <span class="p">|</span> sort <span class="p">|</span> uniq -c
</code></pre></div>
<p>gives:</p>
<div class="highlight"><pre><span></span><code>407 PASS
2 SOR3
</code></pre></div>
<p>So two records were filtered out, based on the Symmetric Odds Ratio (issues with strand bias).</p>
<p>Your script to evaluate the concordance:</p>
<div class="highlight"><span class="filename">evaluate_concordance.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span> ~/workdir/results

gatk Concordance <span class="se">\</span>
--evaluation variants/mother.trio.filtered.vcf <span class="se">\</span>
--truth data/variants/NA12878.vcf.gz <span class="se">\</span>
--intervals chr20:10018000-10220000 <span class="se">\</span>
--summary variants/concordance.mother.trio.filtered
</code></pre></div>
<p>Check out the output with <code>cat</code>:</p>
<div class="highlight"><pre><span></span><code>cat variants/concordance.mother.trio.filtered
</code></pre></div>
<p>gives:</p>
<div class="highlight"><pre><span></span><code>type    TP      FP      FN      RECALL  PRECISION
SNP     319     5       9       0.973   0.985
INDEL   63      20      6       0.913   0.759
</code></pre></div>
<p>Showing that there were 9 false negatives, i.e. SNPs we didn&rsquo;t detect.</p>
</details>
<div class="admonition tip">
<p class="admonition-title">Recall &amp; precision</p>
<p>More info on the definition of recall and precision on this <a href="https://en.wikipedia.org/wiki/Precision_and_recall">wikipedia page</a></p>
</div>
<p><strong>Exercise:</strong> Check out the concordance of the mother with the truth set before filtering. Do this by generating two scripts:</p>
<ul>
<li><code>C16_extract_mother_before_filtering.sh</code>: to run <code>gatk SelectVariants</code> in order to get only variants from the mother from the unfiltered <code>trio.vcf</code>. </li>
<li><code>C17_evaluate_concordance_before_filtering.sh</code>: to run <code>gatk Concordance</code> on the selected variants. </li>
</ul>
<p>Did filtering improve the recall or precision?</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>We did the filtering on <code>trio.vcf</code>, therefore, you first have to extract the records that only apply to the mother by using <code>gatk SelectVariants</code>.</p>
<p>Also note that <code>trio.vcf</code> contains records other than SNPs and INDELs. Use <code>--select-type-to-include</code> to select only SNPs and INDELs.</p>
</div>
<details class="done"><summary>Answer</summary><p>First select only SNPs and INDELs from the mother from the unfiltered vcf:</p>
<div class="highlight"><span class="filename">C16_extract_mother_before_filtering.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span> ~/workdir/results/variants

gatk SelectVariants <span class="se">\</span>
--variant trio.vcf <span class="se">\</span>
--sample-name mother <span class="se">\</span>
--exclude-non-variants <span class="se">\</span>
--remove-unused-alternates <span class="se">\</span>
--select-type-to-include INDEL <span class="se">\</span>
--select-type-to-include SNP <span class="se">\</span>
--output mother.trio.vcf
</code></pre></div>
<p>Get the concordance with the truth set:</p>
<div class="highlight"><span class="filename">C17_evaluate_concordance_before_filtering.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span> ~/workdir/results

gatk Concordance <span class="se">\</span>
--evaluation variants/mother.trio.vcf <span class="se">\</span>
--truth data/variants/NA12878.vcf.gz <span class="se">\</span>
--intervals chr20:10018000-10220000 <span class="se">\</span>
--summary variants/concordance.mother.trio
</code></pre></div>
<p>Which gives:</p>
<div class="highlight"><pre><span></span><code>type    TP      FP      FN      RECALL  PRECISION
SNP     319     7       9       0.973   0.979
INDEL   63      20      6       0.913   0.759
</code></pre></div>
<p>The precision for SNPs is slightly lower. Due to filtering, we removed two false positives.</p>
</details>

              
            </article>
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../visualisation/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Visualisation" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Visualisation
            </div>
          </div>
        </a>
      
      
        
        <a href="../annotation/" class="md-footer__link md-footer__link--next" aria-label="Next: Annotation" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Annotation
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2021 SIB Swiss institute of Bioinformatics
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../..", "features": [], "search": "../../assets/javascripts/workers/search.5e67fbfe.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.c44cc438.min.js"></script>
      
    
  </body>
</html>